### PR TITLE
Add random partitioning on SageMaker. Update run and compose scripts along with instructions.

### PR DESCRIPTION
**Pull request by thvasilo**
*Description of changes:*
Add random partitioning on SageMaker, and support for partitioning using Docker compose.

* The SM Dockerfile is changed to clone DGL as we use `tools/data_dispatch.py` and install all the necessary dependencies.
* We change the way we generate Docker compose files, using a sub-parser with an action for each of `partitioning`, `training`, `inference`. This helps create better separation and logical grouping of arguments.
* We add a general/extensible way to run distributed partition algorithms on SageMaker, with initial support for random partitioning.

Most of the code changes add the new code for partitioning. I would suggest focusing on changed code rather than the new files to see the new way to run `generate_sagemaker_docker_compose.py`, that is e.g.:

```bash
GSF_ARGS="--num-layers 1 --hidden-size 128 --backend gloo --batch-size 128"
python local/generate_sagemaker_docker_compose.py --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    training \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

Most of the changes in the README are about explaining the changes to Docker compose.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        thread.start()
+
+        data_dispatch_step(local_partition_path)
+
+        # Indicate we can stop sending keepalive messages
+        task_end.set()
+        # Ensure the keepalive thread has finished before closing sockets
+        thread.join()
+        # Close connections with workers
+        utils.terminate_workers(client_list, world_size, task_end)
```
Comment (thvasilo) [inline python/graphstorm/sagemaker/sagemaker_partition.py:361]: Rest of executions scripts need to do the same.

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+from graphstorm.sagemaker import utils
+from .s3_utils import download_data_from_s3, upload_file_to_s3
+from .partition_algorithm import RandomPartitioner, PartitionerConfig
+
+DGL_TOOL_PATH = "/root/dgl/tools"
+
+@dataclass
+class PartitionJobConfig():
+    """
+    Configuration object for a SageMaker partitioning job.
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_partition.py:41]: Can you have a description of each field here?

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+
+    # download edge info
+    edges = graph_config["edges"]
+    for etype, edge_data in edges.items():
+        if local_rank == 0:
+            logging.info("Downloading edge structure for edge type '%s'", etype)
+        edge_file_list = edge_data["data"]
+        read_list = np.array_split(np.arange(len(edge_file_list)), world_size)
+        for i, efile in enumerate(edge_file_list):
+            # TODO: Only download round-robin if ParMETIS will run, skip otherwise
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_partition.py:128]: What about random partition?

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+            # Download files both in round robin and sequential assignment
+            if i % world_size == local_rank or i in read_list[local_rank]:
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_partition.py:130]: Why both in round robin and sequential?
It is weird here.

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                graph_config_s3_path, meta_info_file)
+        download_data_from_s3(graph_config_s3_path, tmp_data_path,
+            sagemaker_session=sagemaker_session)
+
+    with open(meta_info_file, 'r') as f: # pylint: disable=unspecified-encoding
+        graph_config = json.load(f)
+
+    logging.info("Downloading graph data from %s into %s",
+        graph_data_s3, tmp_data_path)
+    graph_data_path = download_graph(
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_partition.py:275]: if skip_partitioning, do we need to download the graph data?

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        # All workers + leader need to download partition data from S3
+        s3_partition_path = os.path.join(output_s3, "partition")
+        local_partition_path = os.path.join(graph_data_path, "partition")
+        logging.warning("Skipped partitioning step, trying to download partition data from %s",
+            s3_partition_path)
+
+        download_data_from_s3(s3_partition_path, local_partition_path, sagemaker_session)
+    else:
+        # Perform the partitioning. Leader uploads partition assignments to S3.
+        local_partition_path, s3_partition_path = partitioner.create_partitions(
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_partition.py:305]: So now only leader do the random partition?

File: python/graphstorm/sagemaker/sagemaker_train.py [hunk 7]
```diff
@@ -150,8 +150,8 @@ def run_train(args, unknownargs):
     # start the ssh server
     subprocess.run(["service", "ssh", "start"], check=True)
 
-    print(f"Know args {args}")
-    print(f"Unknow args {unknownargs}")
+    print(f"Known args {args}")
```
Comment (classicsong) [inline python/graphstorm/sagemaker/sagemaker_train.py:153]: Put these into a typo PR.

File: python/graphstorm/sagemaker/utils.py [hunk 8]
```diff
@@ -206,16 +233,40 @@ def download_graph(graph_data_s3, graph_name, part_id, local_path, sagemaker_ses
...
+
+    graph_data_s3 = graph_data_s3[:-1] if graph_data_s3.endswith('/') else graph_data_s3
+
+    # We split on '/' to get the bucket, as it's always the third split element in an S3 URI
+    s3_input_bucket = graph_data_s3.split("/")[2]
+    # Similarly, by having maxsplit=3 we get the S3 key value as the fourth element
+    s3_input_key = graph_data_s3.split("/", maxsplit=3)[3]
+
+    s3_client = boto3.client('s3')
+    for graph_config  in [f"{graph_name}.json", "metadata.json"]:
```
Comment (classicsong) [inline python/graphstorm/sagemaker/utils.py:251]: Why f"{graph_name}.json" and "metadata.json" here?

File: sagemaker/launch/launch_infer.py [hunk 9]
```diff
@@ -109,48 +108,58 @@ def run_job(input_args, image, unknowargs):
...
     parser.add_argument("--task-type", type=str,
+        required=True,
         help=f"Task type in {SUPPORTED_TASKS}")
     parser.add_argument("--yaml-s3", type=str,
+        required=True,
         help="S3 location of inference yaml file. "
              "Do not store it with partitioned graph")
     parser.add_argument("--model-artifact-s3", type=str,
-        help="S3 bucket to load the saved model artifacts")
+        required=True,
```
Comment (classicsong) [inline sagemaker/launch/launch_infer.py:161]: Let's put it into another PR.

File: sagemaker/launch/launch_train.py [hunk 10]
```diff
@@ -65,7 +64,7 @@ def run_job(input_args, image, unknowargs):
 
     container_image_uri = image
 
-    prefix = "script-mode-container"
+    job_prefix = input_args.task_name if input_args.task_name else f"train-{input_args.graph_name}"
```
Comment (classicsong) [inline sagemaker/launch/launch_train.py:67]: Put it into another PR.

File: sagemaker/run/infer_entry.py [hunk 11]
```diff
@@ -21,21 +21,29 @@
 from graphstorm.config import SUPPORTED_TASKS
 from graphstorm.sagemaker.sagemaker_infer import run_infer
 
-def parse_train_args():
-    """  Add arguments for model training
+def get_inference_parser():
```
Comment (classicsong) [inline sagemaker/run/infer_entry.py:24]: Put it into another PR.

File: docker/sagemaker/Dockerfile.sm [hunk 12]
```diff
@@ -13,6 +13,11 @@ RUN pip3 install dgl==1.0.2 -f https://data.dgl.ai/wheels/cu117/repo.html
 # Install related Python packages
 RUN pip3 install ogb==1.3.6 scipy pyarrow boto3 scikit-learn transformers
 
+# Install MPI etc needed by DistDGL
+RUN apt-get update; apt-get install -y --no-install-recommends libopenmpi-dev \
+    build-essential make g++ software-properties-common; add-apt-repository ppa:ubuntu-toolchain-r/test; \
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply docker/sagemaker/Dockerfile.sm:18]: Good point, I took most of the deps from an image where I was building DGL. Let me try to minimize these.

File: python/graphstorm/sagemaker/partition_algorithm.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+class RandomPartitioner(PartitionAlgorithm): # pylint: disable=too-few-public-methods
+    """
+    Single-instance random partitioning algorithm.
+    """
+    def _run_partitioning(self, num_partitions: int) -> str:
+        partition_dir = os.path.join(self.local_output_path, "partition")
+        os.makedirs(partition_dir, exist_ok=True)
+
+        # Random partitioning is done on the leader node only
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply python/graphstorm/sagemaker/partition_algorithm.py:181]: The base class is written assuming distributed execution (e.g. ParMETIS). Here we only utilize the leader to do the partitioning assignments, but the caller and synchronization between stages still happens using the leader/worker scheme.

File: python/graphstorm/sagemaker/partition_algorithm.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+            logging.info("Generating random partition for node type %s", ntype)
+            ntype_output = os.path.join(partition_dir, f"{ntype}.txt")
+
+            partition_assignment = np.random.randint(0, num_partitions, (num_nodes_for_type,))
+
+            arrow_partitions = pa.Table.from_arrays(
+                [pa.array(partition_assignment, type=pa.int64())],
+                names=["partition_id"])
+            options = pa_csv.WriteOptions(include_header=False, delimiter=' ')
+            pa_csv.write_csv(arrow_partitions, ntype_output, write_options=options)
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply python/graphstorm/sagemaker/partition_algorithm.py:198]: It's the expected input for DistDGL's `dispatch_data.py`

File: python/graphstorm/sagemaker/sagemaker_partition.py (hunk 4 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/sagemaker_partition.py:130]: ParMETIS assumes round-robin assignment, while data_dispatch assume sequential. So we need to have both available locally. We could remove the round-robin for now since we only have random though.

File: python/graphstorm/sagemaker/sagemaker_partition.py (hunk 5 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/sagemaker_partition.py:275]: Yes, the local data are used by data_dispatch to create the DistDGL files right?

File: python/graphstorm/sagemaker/sagemaker_partition.py (hunk 6 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/sagemaker_partition.py:305]: For random partitioning yes we don't need to involve the workers as the partitioning takes seconds even for very large graphs. If we run into any performance issues we can revisit.

To be clear: `sagemaker_partition.py` is agnostic of the partitioning algorithm implementation, as long as it follows the contracts of the `PartitionAlgorithm` base class. 

One of the contracts is that at the end of the partition assignment step, the partition assignment files exist on the leader, which proceeds to upload them to the designated location.

File: python/graphstorm/sagemaker/utils.py (hunk 8 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/utils.py:251]: data_dispatch.py creates a file named `metadata.json`. I guess the single-instance partitioning instead creates `{graph_name}.json` so we should support both.

File: python/graphstorm/sagemaker/partition_algorithm.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        world_size: int
+            Size of the distributed training/inference cluster
+        success: bool
+            True if preprocess success
+        """
+        if self.rank != 0:
+            raise RuntimeError("broadcast_partition_done should only be called by the leader")
+        msg = b"PartitionDone" if success else b"PartitionFail"
+        for rank in range(1, world_size):
+            client_list[rank].sendall(msg)
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply python/graphstorm/sagemaker/partition_algorithm.py:117]: `sendall` is safer here right? It ensures the whole buffer is sent before returning.

File: python/graphstorm/sagemaker/s3_utils.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    Parameters
+    ----------
+    input_s3: str
+        S3 uri of the input file
+    local_data_path: str
+        Local file path to store the downloaded file.
+    sagemaker_session: sagemaker.session.Session
+        sagemaker_session to run download
+    """
+    logging.debug("download %s into %s", input_s3, local_data_path)
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply python/graphstorm/sagemaker/s3_utils.py:23]: I think we can leave it as the responsibility of the caller. If you think we need to take care of this I can add a simple check.

**Comment by thvasilo** [comment]
Hi @classicsong I've addressed most of the comments here and resolved the conversations that I think were answered. If you think we need some more changes feel free to re-open one of the resolved comments and I'll address it.

File: sagemaker/local/generate_sagemaker_docker_compose.py [hunk 17]
```diff
@@ -18,111 +18,224 @@
 """
```
Comment (classicsong) [inline sagemaker/local/generate_sagemaker_docker_compose.py:18]: Shall we have separate compose.py for partition and training/inference?

Comment (thvasilo) [reply sagemaker/local/generate_sagemaker_docker_compose.py:18]: My proposed design and implementation here uses one compose file, with different sub-parsers and commands. For example to generate a `training` compose you'd use:


```bash
python local/generate_sagemaker_docker_compose.py --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    training \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

See the README for details.

File: python/graphstorm/sagemaker/sagemaker_partition.py [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+        for ntype, ndata in node_data.items():
+            for feat_name, feat_data in ndata.items():
+                if local_rank == 0:
+                    logging.info("Downloading node feature '%s' of node type '%s'",
+                    feat_name, ntype)
+                num_files = len(feat_data["data"])
+                # TODO: Use dgl.tools.distpartitioning.utils.generate_read_list
+                # once we move the code over from DGL
+                read_list = np.array_split(np.arange(num_files), world_size)
+                for i in read_list[local_rank].tolist():
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply python/graphstorm/sagemaker/sagemaker_partition.py:169]: This arrangement is sequential, e.g.

```python
In [188]: num_files = 8

In [189]: world_size = 4

In [190]: np.array_split(np.arange(num_files), world_size)
Out[190]: [array([0, 1]), array([2, 3]), array([4, 5]), array([6, 7])]
```

File: docker/sagemaker/Dockerfile.sm [hunk 19]
```diff
@@ -24,6 +29,13 @@ ENV PYTHONPATH="/opt/ml/code/graphstorm/python/:${PYTHONPATH}"
 
 RUN cp /opt/ml/code/graphstorm/sagemaker/run/* /opt/ml/code/
 
+# Download DGL source code
+RUN cd /root; git clone https://github.com/dmlc/dgl.git
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply docker/sagemaker/Dockerfile.sm:33]: Yeah, let me do that and add it.

File: python/graphstorm/sagemaker/sagemaker_partition.py (hunk 4 above)
Comment (thvasilo) [reply python/graphstorm/sagemaker/sagemaker_partition.py:147]: I'll have a follow-up PR that only downloads in the way that data_dispatch expects, would that cover it?

File: python/graphstorm/sagemaker/sagemaker_partition.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/sagemaker/sagemaker_partition.py:169]: I c.

File: sagemaker/local/generate_sagemaker_docker_compose.py (hunk 17 above)
Comment (classicsong) [reply sagemaker/local/generate_sagemaker_docker_compose.py:18]: Can we do the following?
```
python local/generate_sagemaker_docker_compose.py  training \
    --image graphstorm:sm  \
    --graph-name ${GRAPH_NAME} --num-instances ${NUM_INSTANCES} \
    --graph-data-s3 ${DATASET_S3_PATH} --region ${REGION} \
    --task-type ${TASK_TYPE} --train-yaml-s3 ${TRAIN_YAML_S3} \
    --model-artifact-s3 ${OUTPUT_S3_PATH} $GSF_ARGS
```

Comment (thvasilo) [reply sagemaker/local/generate_sagemaker_docker_compose.py:18]: Done in latest commit.
